src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -0,0 +1,322 @@
+        /// Append data
+        /// </summary>
+        /// <param name="data">Data</param>
+        /// <param name="size">Size</param>
+        public void AppendData(byte[] data, int size)
+        {
+            Buffer.BlockCopy(data, 0, _data, _length, size);
+            _length += size;
+        }
+
+        /// <summary>
+        /// Append another packet
+        /// </summary>
+        /// <param name="packet">Packet</param>
+        public void AppendPacket(SNIPacket packet)
+        {
+            Buffer.BlockCopy(packet._data, 0, _data, _length, packet._length);
+            _length += packet._length;
+        }
+
+        /// <summary>
+        /// Take data from packet and advance offset
+        /// </summary>
+        /// <param name="buffer">Buffer</param>
+        /// <param name="dataOffset">Data offset</param>
+        /// <param name="size">Size</param>
+        /// <returns></returns>
+        public int TakeData(byte[] buffer, int dataOffset, int size)
+        {
+            if (_offset >= _length)
+            {
+                return 0;
+            }
+
+            if (_offset + size > _length)
+            {
+                size = _length - _offset;
+            }
+
+            Buffer.BlockCopy(_data, _offset, buffer, dataOffset, size);
+            _offset = _offset + size;
+            return size;
+        }
+
+        /// <summary>
+        /// Release packet
+        /// </summary>
+        public void Release()
+        {
+            _length = 0;
+            _capacity = 0;
+            _data = null;
+        }
+
+        /// <summary>
+        /// Reset packet 
+        /// </summary>
+        public void Reset()
+        {
+            _length = 0;
+            _data = new byte[_capacity];
+        }
+
+        /// <summary>
+        /// Read data from a stream asynchronously
+        /// </summary>
+        /// <param name="stream">Stream to read from</param>
+        /// <param name="callback">Completion callback</param>
+        public void ReadFromStreamAsync(Stream stream, SNIAsyncCallback callback)
+        {
+            bool error = false;
+
+            stream.ReadAsync(_data, 0, _capacity).ContinueWith(t =>
+            {
+                Exception e = t.Exception != null ? t.Exception.InnerException : null;
+                if (e != null)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, 0, e.Message);
+                    error = true;
+                }
+                else
+                {
+                    _length = t.Result;
+
+                    if (_length == 0)
+                    {
+                        SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, 0, "Connection was terminated");
+                        error = true;
+                    }
+                }
+
+                callback(this, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
+            });
```

Comment:
This should probably explicitly specify ```, CancellationToken.None, TaskContinuationOptions.DenyChildAttach, TaskScheduler.Default``` unless you really want to be scheduling to the current scheduler.  

Separately, you could then also improve perf a bit by passing object state into the lambda and using that to provide the arguments to the function rather than using closed over state.